【Windows编译错误】1.8版本Windows下/MD编译GPU预测库出错

- 版本、环境信息：
    1）PaddlePaddle版本：1.8，COMMIT ID: 0231f58e592ad9f673ac1832d8c495c8ed65d24f
    2）CPU：i5-9400F, MKL=OFF，使用OPENBLAS
    3）GPU：2080Ti, CUDA10.1，CUDNN7.6
    4）系统环境：Windows10，VS2015专业版
- 安装方式信息：
2）本地编译：请提供cmake命令，编译命令    
```
cmake .. -G "Visual Studio 14 2015" -A x64 -T host=x64 -DCMAKE_BUILD_TYPE=Release -DWITH_MKL=OFF -DWITH_GPU=ON -DON_INFER=ON -DWITH_PYTHON=OFF -DCUDA_TOOKIT_ROOT_DIR="C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v10.1\lib\x64" -DCUDA_ARCH_NAME=All -DMSVC_STATIC_CRT=OFF
```
- 问题描述：
使用 /MD 模式进行编译的话，会在最终出错，出错信息有以下几类：
```
- libprofiler.lib(profiler_generated_profiler.cu.obj) : error LNK2038: 检测到“RuntimeLibrary”的不匹配项: 值“MT_StaticRelease”不匹配值“MD_DynamicRelease”(io.obj 中)
- msvcprt.lib(MSVCP140.dll) : error LNK2005: "public: __cdecl std::locale::id::operator unsigned __int64(void)" (??Bid@locale@std@@QEAA_KXZ) 已经在 libprofiler.lib(profiler_generated_profiler.cu.obj) 中定义
- libcpmt.lib(ios.obj) : error LNK2038: 检测到“RuntimeLibrary”的不匹配项: 值“MT_StaticRelease”不匹配值“MD_DynamicRelease”(io.obj 中)
- libtensor.lib(tensor_generated_.tensor_util.cu.obj) : error LNK2038: 检测到“RuntimeLibrary”的不匹配项: 值“MT_StaticRelease”不匹配值“MD_DynamicRelease”(io.obj 中)
- libtensor.lib(tensor_generated_.tensor_util.cu.obj) : error LNK2005: "public: class std::basic_ostream<char,struct std::char_traits<char> > & __cdecl std::basic_ostream<char,struct std::char_traits<char> >::operator<<(short)" (??6?$basic_ostream@DU?$char_traits@D@std@@@std@@QEAAAEAV01@F@Z) 已经在 msvcprt.lib(MSVCP140.dll) 中定义
- libpaddle_fluid.lib(profiler_generated_profiler.cu.obj) : error LNK2038: 检测到“RuntimeLibrary”的不匹配项: 值“MT_StaticRelease”不匹配值“MD_DynamicRelease”(pd_config.obj 中)
```

--- a/cmake/cuda.cmake
+++ b/cmake/cuda.cmake
@@ -205,9 +205,9 @@ set(CUDA_PROPAGATE_HOST_FLAGS OFF)
 # Release/Debug flags set by cmake. Such as -O3 -g -DNDEBUG etc.
 # So, don't set these flags here.
 if (NOT WIN32) # windows msvc2015 support c++11 natively. 
-# -std=c++11 -fPIC not recoginize by msvc, -Xcompiler will be added by cmake.
-list(APPEND CUDA_NVCC_FLAGS "-std=c++11")
-list(APPEND CUDA_NVCC_FLAGS "-Xcompiler -fPIC")
+  # -std=c++11 -fPIC not recoginize by msvc, -Xcompiler will be added by cmake.
+  list(APPEND CUDA_NVCC_FLAGS "-std=c++11")
+  list(APPEND CUDA_NVCC_FLAGS "-Xcompiler -fPIC")
 endif(NOT WIN32)
 
 # in cuda9, suppress cuda warning on eigen 
@@ -229,6 +229,11 @@ if (NOT WIN32)
 else(NOT WIN32)
   list(APPEND CUDA_NVCC_FLAGS  "-Xcompiler \"/wd 4244 /wd 4267 /wd 4819\"")
   list(APPEND CUDA_NVCC_FLAGS  "--compiler-options;/bigobj")
+  if(MSVC_STATIC_CRT)
+    list(APPEND CUDA_NVCC_FLAGS "-Xcompiler" "-MT$<$<CONFIG:Debug>:d>")
+  else()
+    list(APPEND CUDA_NVCC_FLAGS "-Xcompiler" "-MD$<$<CONFIG:Debug>:d>")
+  endif()
   if(CMAKE_BUILD_TYPE  STREQUAL "Debug")
     list(APPEND CUDA_NVCC_FLAGS  "-g -G")
     # match the cl's _ITERATOR_DEBUG_LEVEL

--- a/cmake/flags.cmake
+++ b/cmake/flags.cmake
@@ -217,8 +217,8 @@ foreach(flag ${GPU_COMMON_FLAGS})
 endforeach()
 
 if(WIN32 AND MSVC_STATIC_CRT)
-# windows build turn off warnings.
-safe_set_static_flag()
+    # windows build turn off warnings.
+    safe_set_static_flag()
     foreach(flag_var
         CMAKE_CXX_FLAGS CMAKE_CXX_FLAGS_DEBUG CMAKE_CXX_FLAGS_RELEASE
         CMAKE_CXX_FLAGS_MINSIZEREL CMAKE_CXX_FLAGS_RELWITHDEBINFO

